sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducerAsyncClient.java

```diff
@@ -447,18 +516,56 @@ public String getEventHubName() {
             // Start send span and store updated context
             parentContext.set(tracerProvider.startSpan(finalSharedContext, ProcessKind.SEND));
         }
-
-        return withRetry(getSendLink(batch.getPartitionId())
-            .flatMap(link ->
-                messages.size() == 1
+        if (this.enableIdempotentPartitions) {
+            PartitionPublishingState publishingState = this.getClientPartitionPublishingState(batch.getPartitionId());
+            try {
+                // Ensure only one EventDataBatch of a partition is being sent at a time.
+                publishingState.getSendingSemaphore().acquire();
+            } catch (InterruptedException e) {
+                return monoError(logger, new RuntimeException(e));
+            }
+            return withRetry(
+                getSendLink(batch.getPartitionId())
+                .map(link -> {
+                    int seqNumber = publishingState.getSequenceNumber();
+                    batch.setStartingPublishedSequenceNumber(seqNumber);
+                    for (EventData eventData : batch.getEvents()) {
+                        eventData.getSystemProperties().put(
+                            AmqpMessageConstant.PRODUCER_SEQUENCE_NUMBER_ANNOTATION_NAME.getValue(),
+                            seqNumber);
+                        seqNumber = PartitionPublishingUtils.incrementSequenceNumber(seqNumber);
+                    }
+                    return link;
+                })
+                .flatMap(link -> messages.size() == 1
+                    ? link.send(messages.get(0))
+                    : link.send(messages)),
+                retryOptions.getTryTimeout(), retryPolicy
+            ).publishOn(scheduler)
+                .doOnEach(signal -> {
+                    if (isTracingEnabled) {
+                        tracerProvider.endSpan(parentContext.get(), signal);
+                    }
+                })
+                .then(Mono.fromRunnable(() -> {
```

Comment:
Forgive my lack of Flux, but in the case where there's a failure state (all retries exhausted), you remove the batch sequence number, each of the event sequence numbers, and don't update state right?

Reply:
Good question. `then` won't be reached if the send has an error.